marlowe-dashboard-client/src/Dashboard/State.purs

```diff
@@ -156,9 +161,28 @@ handleAction input@{ currentSlot } UpdateFromStorage = do
       in
         void $ for unfoldedFollowerApps \(plutusAppId /\ contractHistory) -> handleAction input $ UpdateContract plutusAppId contractHistory
 
--- this handler takes the wallet companion app state (which contains MarloweParams of all the contracts)
--- the wallet is interested in, compares it to the existing contracts, and either creates new follower
--- apps or activates pending ones for any contracts that aren't yet being followed
+{- [Workflow 2][6] Connect a wallet
+If we have just connected a wallet (and the walletCompanionStatus is still `FirstUpdatePending`),
+then we know the `MarloweFollower` apps that are running in this wallet, and have just been given
+the current state of its `WalletCompanion` app for the first time since connecting. If the status
+of the `WalletCompanion` app contains a record of `MarloweParams` for any _new_ Marlowe contracts
+(created since the last time we connected this wallet, and for which we therefore have no
+corresponding `MarloweFollowr` apps), we now need to create `MarloweFollower` apps for those new
+contracts.
+In this case, we change the walletCompanionStatus to `LoadingNewContracts`, and subscribe to every
+new `MarloweFollower`. We'll know the loading is finished when we've received the first status
+update for each of these `MarloweFollower` apps through the WebSocket.
+-}
+{- [Workflow 4][2] Start a Marlowe contract
+After starting a new Marlowe contract, we should receive a WebSocket notification informing us of
+the `MarloweParams` and initial `MarloweData` for that contract (via a status update for our
+`WalletCompanion` app). We now need to start following that contract with a `MarloweFollower` app.
+If we started the contract ourselves, we already created a `MarloweFollower` app as a placeholder.
+So here we need to check whether there is an existing `MarloweFollower` app with the right metadata
+(and no `MarloweParams`) - and potentially use that instead of creating a new one.
+If someone else started the contract, and gave us a role, we will have no placeholder
+`MarloweFollower` app, and so we simply create a new one and start following immediately.
+-}
```

Comment:
😬 It's not good that the same action is triggered as part of two separate workflows... Each workflow step should have one action for it. This is what happens when you build an event sourced system with general-purpose CRUD events, rather than events which have meaning in the domain.

We should consider refactoring this at some point - this will get exponentially worse as system complexity grows.

Reply:
yes, the trouble part is that we don't know what the MarloweParams are when we create a contract. Today I suggested that after attaching a requestId, we could modify the state of the MarloweApp contract to move from `Map MarloweParams MarloweData` to `Map UUID (Tuple MarloweParams MarloweData)` so that we can later tie a request to create a contract with what id that contract is gonna have.

Also, we definitely have to improve the names `MarloweParams` and `MarloweData`